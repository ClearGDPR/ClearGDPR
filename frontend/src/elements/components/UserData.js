import React from 'react';
import Elements from '../components/Elements';
import styles from '../theme/SubjectData.scss';

// Check if OG is initialized
Elements();

class UserData extends React.PureComponent {
  state = {
    data: null,
    status: null
  };

  componentDidMount() {
    // TODO: Improve cache, save to elements store, add localStorage
    setTimeout(() => {
      const cgToken = localStorage.getItem('cgToken');

      window.cg.setAccessToken(cgToken);
      window.cg.Subject.accessData()
        .then(data => {
          this.setState({ data });
        })
        .catch(err => {
          console.log('failure', err);
        });

      window.cg.Subject.getDataStatus()
        .then(status => {
          this.setState({ status });
        })
        .catch(err => {
          console.log('failure', err);
        });
    }, 1000);
  }

  render() {
    const { data, status } = this.state;

    if (!data || !status) {
      return <div>Loading</div>;
    }

    return (
      <div className={styles.data}>
        <b>Data shared: </b>
        <br />
        <ul>
          {Object.keys(data).map(field => (
            <li key={field}>
              <b>{field}</b>: {data[field]}
            </li>
          ))}
        </ul>
        <br />
        <b>Your data is shared with: </b>
        <div>Controller: {status.controller}</div>
        <br />
        <b>Processors</b>
        <ul>
          {status.processors.map(p => (
            <div key={p.id} className={styles.processor}>
              <span
                className={[styles.status, p.status ? styles.isActive : styles.isInactive].join(
                  ' '
                )}
              />
              <span>{p.id}</span>
            </div>
          ))}
        </ul>
      </div>
    );
  }
}

export default UserData;
