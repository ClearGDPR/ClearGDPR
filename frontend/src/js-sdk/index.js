/*
 * cg-js-sdk
 * https://www.cleargdpr.com/
 * 
 */
export { CG } from './cg';
