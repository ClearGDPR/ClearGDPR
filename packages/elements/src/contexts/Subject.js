import _ from 'lodash';
import { DATA_ERASED, DATA_STATUS } from '../constants';

export default class Subject {
  constructor(
    cg,
    {
      isFetched = false,
      isLoading = false,
      isErased = false,
      data = null,
      status = null,
      processors = null,
      objection = null,
      shares = null,
      restrictions = null,
      isGuest = true,
      propagateMutation
    }
  ) {
    this.cg = cg;

    this.isFetched = isFetched;
    this.isLoading = isLoading;
    this.isErased = isErased;
    this.data = data;
    this.status = status;
    this.processors = processors;
    this.objection = objection;
    this.shares = shares;
    this.restrictions = restrictions;
    this.isGuest = isGuest;

    this._propagateMutation = () => {
      propagateMutation(new Subject(cg, { ...this, propagateMutation }));
    };
  }

  async fetchData() {
    const { isLoading, isFetched, isGuest } = this;

    if (isLoading || isFetched || isGuest) {
      return;
    }

    this.isLoading = true;

    let data;
    let status;

    try {
      status = await this.cg.Subject.getDataStatus();
    } catch (e) {
      status = null;
    }

    const isErased = !status || status.controller === DATA_ERASED;

    if (!isErased) {
      try {
        data = await this.cg.Subject.accessData();
      } catch (e) {
        data = null;
      }
    }

    this.data = data;
    this.isLoading = false;
    this.isFetched = true;
    this.isErased = isErased;
    this.status = status;

    this._propagateMutation();
  }

  async fetchProcessors() {
    if (this.processors) {
      return;
    }

    let processors;
    try {
      processors = await this.cg.Subject.getProcessors();
    } catch (e) {
      console.log('failure', e);
      return;
    }

    this.processors = processors.map(p => {
      p.enabled = true;

      if (this.status) {
        const s = _.find(this.status.processors, { id: p.id }) || {};
        p.status = s.status ? DATA_STATUS[s.status] : DATA_STATUS.UNCONSENTED;
      }

      return p;
    });

    this._propagateMutation();
  }

  async initiateRectification(requestReason, rectificationPayload) {
    return await this.cg.Subject.initiateRectification(requestReason, rectificationPayload);
  }

  async eraseData() {
    await this.cg.Subject.eraseData();
    this.status.controller = DATA_ERASED;
    this.data = null;
    this.isErased = true;
    this._propagateMutation();
  }

  async fetchObjectionStatus() {
    if (this.objection !== null) {
      return;
    }
    const { objection } = await this.cg.Subject.getObjectionStatus();
    this.objection = objection;
    this._propagateMutation();
  }

  async updateObjection(value) {
    await this.cg.Subject.updateObjection(value);
    this.objection = value;
    this._propagateMutation();
  }

  async fetchDataShares() {
    if (this.shares !== null) {
      return;
    }
    try {
      this.shares = await this.cg.Subject.getDataShares();
    } catch (e) {
      console.log('failure', e);
      return;
    }

    this._propagateMutation();
  }

  async addDataShare(name) {
    try {
      await this.cg.Subject.addDataShare({
        name
      });
    } catch (e) {
      console.log(e);
      return;
    }
    this.shares = null;
    this.fetchDataShares();
  }

  async removeDataShare(id) {
    try {
      await this.cg.Subject.removeDataShare(id);
    } catch (e) {
      console.log(e);
      return;
    }
    this.shares = null;
    this.fetchDataShares();
  }

  async fetchRestrictions() {
    if (this.restrictions !== null) {
      return;
    }
    this.restrictions = await this.cg.Subject.getRestrictions();
    this._propagateMutation();
  }

  async updateRestrictions(restrictions) {
    this.restrictions = restrictions;
    this._propagateMutation();

    try {
      await this.cg.Subject.updateRestrictions(restrictions);
    } catch (e) {
      console.log(e);
    }
  }

  async giveConsent(payload) {
    try {
      await this.cg.Subject.giveConsent(payload);
    } catch (e) {
      console.log('failure', e);
    }
  }

  onEntrance(callback) {
    this.cg.Events.subscribe('auth.setAccessToken', callback);
  }

  clearEntranceListeners() {
    this.cg.Events.clear('auth.setAccessToken');
  }
}
