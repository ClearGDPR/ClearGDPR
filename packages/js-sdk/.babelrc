{
  "presets": [
    [
      "env",
      {
        "targets": {
          "browsers": ["> 1%", "last 2 versions"]
        }
      }
    ]
  ],
  "plugins": [
    [
      "module-resolver",
      {
        "extensions": [".js", ".jsx", ".es", ".es6", ".mjs"],
        "root": ["./src"],
        "alias": {}
      }
    ]
  ]
}
