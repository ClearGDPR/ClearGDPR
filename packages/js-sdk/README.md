<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**  *generated with [DocToc](https://github.com/thlorenz/doctoc)*

- [JS-SDK](#js-sdk)
  - [Related reads](#related-reads)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

# JS-SDK

SDK design is based on several patterns, but main idea is to provide a way to abstract calls to the CG API, and have a way to trigger internal Events when some specific actions is done.

```
├─ config
├─ scripts
└─ src
    ├─ common     # Reusable functions
    ├─ factories  # Service provider of different functions
    ├─ modules    # Each module represents a specification of the CG API
    ├─ cg.js      # CG SDK definition require Token and register all the modules
    ├─ events.js  # Events subscription/notification used by different flows
    ├─ config.js  # Default configuration, could be overwritten
    └─ index.js
```

## Related reads

- https://github.com/hueitan/javascript-sdk-design
- https://github.com/dropbox/dropbox-sdk-js
- https://developers.facebook.com/docs/javascript/reference/v3.1
