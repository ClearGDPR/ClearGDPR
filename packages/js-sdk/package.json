{
  "name": "@cleargdpr/js-sdk",
  "description": "ClearGDPR JS SDK",
  "version": "1.0.0",
  "scripts": {
    "build": "node scripts/build.js",
    "postinstall": "node scripts/build.js",
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "dependencies": {
    "babel-core": "6.25.0",
    "babel-eslint": "7.2.3",
    "babel-jest": "20.0.3",
    "babel-loader": "7.1.1",
    "babel-runtime": "6.26.0",
    "chalk": "1.1.3",
    "es6-promises": "^1.0.10",
    "eslint": "4.4.1",
    "eslint-config-prettier": "^2.8.0",
    "eslint-loader": "1.9.0",
    "eslint-plugin-flowtype": "2.35.0",
    "eslint-plugin-import": "2.7.0",
    "eslint-plugin-jsx-a11y": "5.1.1",
    "eslint-plugin-prettier": "^2.4.0",
    "extract-text-webpack-plugin": "3.0.0",
    "file-loader": "0.11.2",
    "file-saver": "^1.3.8",
    "fs-extra": "3.0.1",
    "isomorphic-fetch": "^2.2.1",
    "jest": "20.0.4",
    "lodash": "^4.17.10",
    "object-assign": "4.1.1",
    "prettier": "^1.8.2",
    "promise": "8.0.1",
    "raf": "^3.4.0",
    "react-dev-utils": "^4.1.0",
    "webpack": "^4.17.1",
    "webpack-manifest-plugin": "1.2.1",
    "whatwg-fetch": "2.0.3"
  },
  "devDependencies": {
    "babel-plugin-module-resolver": "3.1.1",
    "eslint-import-resolver-babel-module": "4.0.0",
    "eslint-plugin-jest": "^21.17.0",
    "uglifyjs-webpack-plugin": "^1.3.0"
  }
}
