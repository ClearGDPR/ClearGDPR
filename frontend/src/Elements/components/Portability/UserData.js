import React from 'react';
import PropTypes from 'prop-types';

import styles from 'theme/SubjectData.scss';
import Subject from 'contexts/Subject';

class UserData extends React.PureComponent {
  componentDidMount() {
    this.props.subject.fetchProcessors();
  }

  render() {
    const { data, status, processors } = this.props.subject;

    if (!data || !status || !processors) {
      return <div>Loading</div>;
    }

    return (
      <div className={styles.data}>
        <b>Data shared: </b>
        <br />
        <ul>
          {Object.keys(data).map(field => (
            <li key={field}>
              <b>{field}</b>: {data[field]}
            </li>
          ))}
        </ul>
        <br />
        <b>Your data is shared with: </b>
        <div>Controller: {status.controller}</div>
        <br />
        <b>Processors</b>
        {processors.map(p => (
          <div key={p.id} className={styles.processor}>
            <div>
              <img src={p.logoUrl} alt={p.name} style={{ height: 60 }} />
            </div>
            <div>{p.name}</div>
            <div>{p.description}</div>
            <div>
              <b>Status</b>: {p.status}
            </div>
          </div>
        ))}
      </div>
    );
  }
}

UserData.propTypes = {
  subject: PropTypes.instanceOf(Subject)
};

export default UserData;
