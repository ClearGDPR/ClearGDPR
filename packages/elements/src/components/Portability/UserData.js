import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';

import styles from '../../theme/SubjectData.scss';
import Subject from '../../contexts/Subject';

const SUBJECT_DATA_STATUS = ['Unconsented', 'Consented', 'Erased'];

class UserData extends React.PureComponent {
  componentDidMount() {
    this.props.subject.fetchProcessors();
  }

  render() {
    const { data, status, processors } = this.props.subject;

    if (!data || !status) {
      return <div>Loading</div>;
    }

    let processorsList = [];

    if (processors && status.processors) {
      processorsList = processors.map(p => {
        const s = _.find(status.processors, { id: p.id });
        p.status = SUBJECT_DATA_STATUS[s.status];
        return p;
      });
    }

    return (
      <div className={styles.data}>
        <b>Data shared: </b>
        <br />
        <ul>
          {Object.keys(data).map(field => (
            <li key={field}>
              <b>{field}</b>: {data[field]}
            </li>
          ))}
        </ul>
        <br />
        <b>Your data is shared with: </b>
        <div>Controller: {status.controller}</div>
        <br />
        <b>Processors</b>
        {processorsList.map(p => (
          <div key={p.id} className={styles.processor}>
            <div>
              <img src={p.logoUrl} alt={p.name} />
            </div>
            <div>{p.name}</div>
            <div>{p.description}</div>
            <div>
              <b>Status</b>: {p.status}
            </div>
          </div>
        ))}
      </div>
    );
  }
}

UserData.propTypes = {
  subject: PropTypes.instanceOf(Subject)
};

export default UserData;
