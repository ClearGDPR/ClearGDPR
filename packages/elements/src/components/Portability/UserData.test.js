import { CG } from '@cleargdpr/js-sdk';
import { shallow } from 'enzyme/build/index';
import React from 'react';
import UserData from './UserData';
import Subject from '../../contexts/Subject';

const processors = [
  {
    id: 1,
    name: 'Nielsen',
    logoUrl: '',
    consented: 72,
    description: '',
    scopes: ['Full Name', 'Email', 'Phone Number']
  },
  {
    id: 2,
    name: 'Live Ramp',
    logoUrl: '',
    consented: 64,
    description: '',
    scopes: ['Email']
  },
  {
    id: 3,
    name: 'Experian',
    logoUrl: '',
    consented: 25,
    description: '',
    scopes: ['Full Name', 'Phone Number']
  }
];

const cg = new CG({
  apiKey: 'test',
  apiUrl: 'test'
});
cg.Subject.getProcessors = async () => {
  return processors;
};

const setup = async () => {
  return shallow(
    <UserData
      {...{
        subject: new Subject(cg, {
          data: { name: 'John', email: 'john@gmail.com' },
          status: 1,
          propagateMutation: () => {}
        })
      }}
    />
  );
};

describe('UserData', () => {
  it('should render correctly', async () => {
    const element = await setup();
    expect(element).toMatchSnapshot();
  });
});
