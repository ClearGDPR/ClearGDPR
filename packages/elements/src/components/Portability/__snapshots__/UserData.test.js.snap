// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserData should render correctly 1`] = `
<div>
  Loading
</div>
`;
